subprojects/core/src/test/groovy/org/gradle/internal/operations/DefaultBuildOperationQueueTest.groovy

```diff
@@ -103,6 +106,59 @@ class DefaultBuildOperationQueueTest extends Specification {
         5    | 10
     }
 
+    def "does not submit more work processors than #threads threads"() {
+        given:
+        setupQueue(threads)
+        lease.leaseFinish() // Release worker lease to allow operation to run, when there is max 1 worker thread
+
+        def expectedWorkerCount = Math.min(runs, threads)
+        def workersStarted = new AtomicInteger()
+        def startedLatch = new CountDownLatch(expectedWorkerCount)
+        def releaseLatch = new CountDownLatch(1)
+        def operationAction = {
+            if (workersStarted.get() < expectedWorkerCount) {
+                println "started worker in thread ${Thread.currentThread().id} (waiting for ${expectedWorkerCount - workersStarted.incrementAndGet()}).."
+            }
+        }
+        def executor = Mock(Executor)
+        def delegateExecutor = Executors.newFixedThreadPool(threads)
+        operationQueue = new DefaultBuildOperationQueue(false, workerRegistry, executor, new SimpleWorker())
+
+        println "expecting ${expectedWorkerCount} concurrent work processors to be started..."
```

Comment:
🧐 Should this be an assertion?

Reply:
This is just an info message to make it clear for troubleshooting purposes when you're looking at the output of a failed test, but the actual assertion of this is down further at line 149.